### PR TITLE
Tab reordering

### PR DESCRIPTION
**Pull request by KaiTheRedNinja**
# Description

Credit to Lingxi Li for most of the tab reordering code.
1. Added tab reordering
2. Fixed the bug where tabs would sometimes not show up

However, this PR has a few bugs.
1. Dragging has that weird effect where the cursor flickers between the text input cursor and the normal cursor
2. When dragging a file tab, the tab contents will get selected

@wdg also may need to do some modifications to WelcomeView, as currently tabContents only allows for a single tab's view to be loaded at a time, meaning that every time tabs are switched, the content in the old tab is erased. I'll make a bug report for that later.

# Related Issue

* #40 

# Checklist

- [x] Fix performance issues with opening new tabs (async stuff everywhere)

- [x] I read and understood the [contributing guide](https://github.com/AuroraEditor/AuroraEditor/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/AuroraEditor/AuroraEditor/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] I documented my code
- [x] Review requested

# Screenshots

### CODE REVIEW

**Comment by nanashili** [comment]
UI Issues with overlapping:
Screenshot 2022-09-12 at 12 01 18

**Comment by nanashili** [comment]
Also the project navigator keeps collapsing

https://user-images.githubusercontent.com/63672227/189627464-b83aa58a-13e5-4797-a638-265bedf0c5a5.mov

**Comment by nanashili** [comment]
It also highlights the opened file when dragging a tab

**Comment by KaiTheRedNinja** [comment]
> It also highlights the opened file when dragging a tab

It did this even before the changes, if you dragged in the tab bar area it would select the text. I think this is a STTextView issue.

**Comment by nanashili** [comment]
> > It also highlights the opened file when dragging a tab
> 
> It did this even before the changes, if you dragged in the tab bar area it would select the text. I think this is a STTextView issue.

Doubt it, because CE uses the same text view as us

File: AuroraEditor/Base/FileSystemClient/Model/FileItem/FileItem.swift [hunk 1]
```diff
@@ -58,9 +58,8 @@ public extension FileSystemClient {
             source.resume()
             self.watcher = source
 
-            // reindex the current item, because the files in the item may have changed
+            // TODO: reindex the current item, because the files in the item may have changed
             // since the initial load on startup.
```
Comment (wdg) [inline AuroraEditor/Base/FileSystemClient/Model/FileItem/FileItem.swift:62]: did you remove `fileSystemClient?.reloadFromWatcher(sourceFileItem: self)` intentionally?

**Review by wdg — APPROVED** [review]

File: AuroraEditor/Base/FileSystemClient/Model/FileItem/FileItem.swift (hunk 1 above)
Comment (nanashili) [reply AuroraEditor/Base/FileSystemClient/Model/FileItem/FileItem.swift:62]: It caused folders to collapse so I think he removed it for that reason

**Comment by wdg** [comment]
> > > It also highlights the opened file when dragging a tab
> > 
> > 
> > It did this even before the changes, if you dragged in the tab bar area it would select the text. I think this is a STTextView issue.
> 
> Doubt it, because CE uses the same text view as us

We use it in an completely different way, so it *can* be different. still no problem.

@KaiTheRedNinja Create a issue for the highlighting, and mention it in this PR, then we can merge it. its not breaking, only annoying.

**Comment by KaiTheRedNinja** [comment]
Once the tests finish I'll merge this. I've already reported the weird bug with text selection.

**Pull request merged**